### ISSUE TITLE
APSFOUND-56: Metrics for JAX-RS

### ISSUE THREAD

**Issue by atosak** [OpenLiberty/open-liberty#8620]
We do not have good metrics for JAX-RS components. We need to get these given the increasing importance of REST.

What commercial Liberty edition does this feature belong in (pick one)?
  Open Liberty

Should this feature be included in the Bluemix buildpack?
 Yes 

What SalesForce Skill/Mission Team and legacy RETAIN queue will be responsible for supporting this feature?

 SalesForce: WAS: L3 JAX-RS
 RETAIN: WL3JRS 

WAD is attached, and available here:[]( https://ibm.box.com/s/j7ytrkejduob5r58echql6f69791f39r​)

​

**Comment by atosak** [comment]
WAD review comments:
My notes, roughly chronological.

-     Cust Overview: image for Alasdair
-     comms: add operations
-     User Stories:  Both ops / SRE roles 
- Stretch items should be lower ranked but explicit user stories. Clarify inbound/outbound granularity in stretch (reported separate vs. not accounted for anywhere)
-  JAXRS METRICS IN MICROPROFILE (XXX: Eric sending a note about where to work this)

        -  maybe non-spec in 1.1 and spec in 2.0 (which will already be disruptive once)
        -   monitor-1.0 autofeature vs. metrics 2.0/2.1 when standardized.o
        -  split out the JMX bits from PMI or re-impl jaxrs metrics
        -  IAN: jaxrs metrics w/ mp-$future w/o monitor feature added

- To-Be: mptracing and stats should cover similar elapsed times.
-  enablement stuff ported back on slide 2 -- "this feature" a little confusing (AHA vs. liberty feature)
-  opentracing allows using URL instead of class/method, so metrics should be able to agree.

        -  don't forget about PATH_INFO or query string stuff that may map to same or different jaxrs resource
        -  opentracing pointer from Don:  To use the operation name format, set a MicroProfile Config version 1.3 compliant configuration key, mp.opentracing.server.operation-name-provider, to http-path. The default valueis class-method.

- re: serviceability Adds additional metrics, no replacing IBMJAXRSERVLET

-     ops cares about URLs and methods more than resources -- user story.  

-   But it may be confusing when PATH_INFO or query string has data

-     http method included in default metric name instead of just java method

-     Feature Design confirm/show response time metric in addition to count in example

-     can HPA figure out RPS or avg response time with our data?

-     serviceability: do something in monitor-1.0 filtering to make the new stuff filterable 

-    SVT: show in prometheus, HPA thing 

-    Once we have data, work with Don on updated grafana dashboard   

-  Mention no relation to requestTiming

**Comment by atosak** [comment]
Andy McCright
| | May 01, 2019

Everything in the list above has been addressed except for:

    Once we have data, work with Don on updated grafana dashboard

​@Jim Krueger and I will work with Don in the next week or so to complete this.

The updated WAD has been uploaded to this feature and to the Box directory.

**Comment by andymc12** [comment]
Link to WAD:  https://ibm.box.com/s/j7ytrkejduob5r58echql6f69791f39r

**Comment by jim-krueger** [comment]
Feature Ship Release Readiness Checklist 
 
Instructions:
- Paste everything below the next line into the GitHub epic for a feature as the first comment.

==================================================================================
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)
 
Instructions:
 - Do the actions below and mark them complete in the checklist when they are done.
 - Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **Before Development Starts or 8 weeks before Onboarding** 
- [ ] POC Design / WAD Review Scheduled (David Chang) or N/A. AHA Feature marked.
- [ ] POC Design / WAD Reviewed (Feature Owner) or N/A. AHA Feature marked.
- [ ] Complete any To-Do's from the POC Review, and create a new TO-DO in the AHA Feature for the "Design Approval Request" and assign it to Ian Robinson. Leave the date blank. (Feature Owner)
- [ ] Design / WAD Approval (Ian Robinson) or N/A. AHA Feature marked.
- [ ] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A. AHA Feature marked.
- [ ] AHA Feature has link to Github epic or issue. Github epic has link back to AHA Feature. (Feature Owner or Backlog Subtribe PM).
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of the epic entitled "FAT Approval Test Summary".  Add and fill in the template as described here: https://github.ibm.com/was-liberty/WS-CD-Open/wiki/Feature-Review-(Feature-Test-Summary-Process)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **3 weeks before Onboarding** 
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** 3 weeks before Onboarding** 
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **2 weeks before Onboarding** 
- [ ] Implementation complete. (Epic owner / Feature owner) 
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approval or No Design Approval on the AHA Feature.
- [ ] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Globalization** (Marika Joannidis - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
- [ ] **Demo** - (Liberty only - Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **1 week before Onboarding** 
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A. 
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **General Availability** 
- [ ] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by jim-krueger** [comment]
Code merged under https://github.com/OpenLiberty/open-liberty/pull/9202
and
https://github.com/OpenLiberty/open-liberty/pull/10450/

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

**Comment by gscottj** [comment]
The output of this feature may be used to generate a user interface in other components or third party offerings. However, this feature does not block or prevent those other components from being enabled to work with assistive technologies.

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below.  You can use either one to create the education.  

Slide Template:  https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Github Template:  https://pages.github.ibm.com/WASL3/site/STE/about

Please upload the completed slides to the same STE Archive BOX folder.  Thanks!

**Comment by chirp1** [comment]
Approving based on the info that I added to the doc issue: https://github.com/OpenLiberty/docs/issues/745

**Comment by rumanaHaque** [comment]
SVT for this was done by Andrew Rieder while testing for MicroProfile Metrics 2.3 Implementation, using issue - https://github.com/OpenLiberty/open-liberty/issues/10914
After adding mpMetrics-2.3 and jaxrsMonitor-1.0 to the application's 'server.xml' file, he accessed a JAX-RS resource, and saw the corresponding jax-rs metrics being displayed.
SVT for this is complete, so approving the SVT record for this issue.

**Comment by jim-krueger** [comment]
Closing issue.  All approvals met.
